Update to ember-source 5.1 types (and other fixes)

leepfrog opened the pull request:
- global.d.ts: Change global type defintions to import ember-source/types and ember-cli-fastboot
- package.json: Remove @types/ember*
- package.json: Update @embroider/addon-dev to 3.2.0
- package.json: Replace babel-eslint with @babel/eslint-parser to address deprecation warning
- package.json: Add @simple-dom/document to import types
- package.json: Add ember-source to import types
- helpers/page-title.ts: Refactor `object` to `Owner` for `constructor()`
- services/page-title.ts: Add guards for potential undefined variables
- services/page-title.ts: Update import to address ember deprecation
- tsconfig.json: Add support for experimental decorators (ie `@service`)

knownasilya commented on addon/dev-only-types/global.d.ts:
```diff
@@ -1,7 +1,2 @@
-  }
-}
+import 'ember-source/types';
+import 'ember-cli-fastboot';
```
❤️

knownasilya commented:
Looks like an install error, I think we can update to node 16 here esp since new addons start at node 16

leepfrog commented:
I think this has another outdated dependency or configuration issue 👀 : 

```shell
$ 
> yarn install
yarn install v1.22.19
[1/4] 🔍  Resolving packages...
success Already up-to-date.
$ cd addon && yarn build
yarn run v1.22.19
$ rollup --config
(!) You have passed an unrecognized option
Unknown output options: experimentalMinChunkSize. Allowed options: amd, assetFileNames, banner, chunkFileNames, compact, dir, dynamicImportFunction, entryFileNames, esModule, exports, extend, externalLiveBindings, file, footer, format, freeze, generatedCode, globals, hoistTransitiveImports, indent, inlineDynamicImports, interop, intro, manualChunks, minifyInternalExports, name, namespaceToStringTag, noConflict, outro, paths, plugins, preferConst, preserveModules, preserveModulesRoot, sanitizeFileName, sourcemap, sourcemapBaseUrl, sourcemapExcludeSources, sourcemapFile, sourcemapPathTransform, strict, systemNullSetters, validate

 → dist...
[!] (plugin Typescript) TypeError: nodes is not iterable
TypeError: nodes is not iterable
    at visitNodes (ember-cli/ember-page-title/node_modules/typescript/lib/typescript.js:27715:26)
    at forEachChildInTypeAliasDeclaration (ember-cli/ember-page-title/node_modules/typescript/lib/typescript.js:28408:98)
    at Object.forEachChild (ember-cli/ember-page-title/node_modules/typescript/lib/typescript.js:27801:37)
    at fixupParentReferences (ember-cli/ember-page-title/node_modules/ts-clone-node/src/clone-node/util/set-parents.ts:8:13)
    at setParents (ember-cli/ember-page-title/node_modules/ts-clone-node/src/clone-node/util/set-parents.ts:31:2)
    at executePreserveNode (ember-cli/ember-page-title/node_modules/ts-clone-node/src/clone-node/clone-node.ts:333:2)
    at Module.preserveNode (ember-cli/ember-page-title/node_modules/ts-clone-node/src/clone-node/clone-node.ts:265:2)
    at preserveMeta (ember-cli/ember-page-title/node_modules/rollup-plugin-ts/src/service/transformer/declaration-bundler/util/clone-node-with-meta.ts:17:9)
    at visitTypeAliasDeclaration (ember-cli/ember-page-title/node_modules/rollup-plugin-ts/src/service/transformer/declaration-bundler/transformers/ensure-no-export-modifier-transformer/visitor/visit-type-alias-declaration.ts:9:9)
    at visitNode (ember-cli/ember-page-title/node_modules/rollup-plugin-ts/src/service/transformer/declaration-bundler/transformers/ensure-no-export-modifier-transformer/visitor/visit-node.ts:27:10)
    at Object.continuation (ember-cli/ember-page-title/node_modules/rollup-plugin-ts/src/service/transformer/declaration-bundler/transformers/ensure-no-export-modifier-transformer/ensure-no-export-modifier-transformer.ts:32:4)
    at ember-cli/ember-page-title/node_modules/rollup-plugin-ts/src/service/transformer/declaration-bundler/transformers/ensure-no-export-modifier-transformer/ensure-no-export-modifier-transformer.ts:39:68
    at visitArrayWorker (ember-cli/ember-page-title/node_modules/typescript/lib/typescript.js:85983:51)
    at visitNodes2 (ember-cli/ember-page-title/node_modules/typescript/lib/typescript.js:85954:21)
    at visitLexicalEnvironment (ember-cli/ember-page-title/node_modules/typescript/lib/typescript.js:86010:18)
    at visitEachChildOfSourceFile (ember-cli/ember-page-title/node_modules/typescript/lib/typescript.js:87199:13)
    at Object.visitEachChild (ember-cli/ember-page-title/node_modules/typescript/lib/typescript.js:86171:35)
    at ensureNoExportModifierTransformer (ember-cli/ember-page-title/node_modules/rollup-plugin-ts/src/service/transformer/declaration-bundler/transformers/ensure-no-export-modifier-transformer/ensure-no-export-modifier-transformer.ts:39:14)
    at applyTransformers (ember-cli/ember-page-title/node_modules/rollup-plugin-ts/src/service/transformer/declaration-bundler/util/apply-transformers.ts:12:31)
    at ember-cli/ember-page-title/node_modules/rollup-plugin-ts/src/service/transformer/declaration-bundler/transformers/source-file-bundler/source-file-bundler.ts:119:32
    at transformation (ember-cli/ember-page-title/node_modules/typescript/lib/typescript.js:109610:16)
    at transformRoot (ember-cli/ember-page-title/node_modules/typescript/lib/typescript.js:109633:73)
    at transformNodes (ember-cli/ember-page-title/node_modules/typescript/lib/typescript.js:109618:72)
    at emitDeclarationFileOrBundle (ember-cli/ember-page-title/node_modules/typescript/lib/typescript.js:110256:36)
    at emitSourceFileOrBundle (ember-cli/ember-page-title/node_modules/typescript/lib/typescript.js:110143:7)
    at forEachEmittedFile (ember-cli/ember-page-title/node_modules/typescript/lib/typescript.js:109886:24)
    at emitFiles (ember-cli/ember-page-title/node_modules/typescript/lib/typescript.js:110114:5)
    at emitWorker (ember-cli/ember-page-title/node_modules/typescript/lib/typescript.js:117550:26)
    at ember-cli/ember-page-title/node_modules/typescript/lib/typescript.js:117535:53
    at runWithCancellationToken (ember-cli/ember-page-title/node_modules/typescript/lib/typescript.js:117624:16)
    at Object.emit (ember-cli/ember-page-title/node_modules/typescript/lib/typescript.js:117535:22)
    at emitNextAffectedFile (ember-cli/ember-page-title/node_modules/typescript/lib/typescript.js:121249:36)
    at Object.emit (ember-cli/ember-page-title/node_modules/typescript/lib/typescript.js:121370:39)
    at runEmit (ember-cli/ember-page-title/node_modules/rollup-plugin-ts/src/service/compiler-host/compiler-host.ts:111:6)
    at CompilerHost.emit (ember-cli/ember-page-title/node_modules/rollup-plugin-ts/src/service/compiler-host/compiler-host.ts:124:3)
    at bundleDeclarationsForChunk (ember-cli/ember-page-title/node_modules/rollup-plugin-ts/src/service/emit/declaration/bundle-declarations-for-chunk.ts:14:34)
    at emitDeclarations (ember-cli/ember-page-title/node_modules/rollup-plugin-ts/src/service/emit/declaration/emit-declarations.ts:174:24)
    at Object.generateBundle (ember-cli/ember-page-title/node_modules/rollup-plugin-ts/src/plugin/typescript-plugin.ts:573:5)
    at ember-cli/ember-page-title/node_modules/rollup/dist/shared/rollup.js:22879:40

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

leepfrog commented:
Ok, I think this is going to work now?  🤞

knownasilya commented:
@leepfrog running..

leepfrog commented:
Ah, tests are failing because the test-app is older ember.

either can:
- update test app to be newer (currently 3.28lts — basically dropping older ember support)
- revert my change and continue to use older import path for `getOwner` (future technical debt due to deprecation)

knownasilya commented:
Lets update the test app

leepfrog commented:
I got stuck on a different issue.  I'm getting an error when the PageTitleService attempts to inject the Router Service (in its constructor):

```
Error: Assertion Failed: Attempting to lookup an injected property on an object without a container, ensure that the object was instantiated via a container.
```

I started [a thread on discord](https://discord.com/channels/480462759797063690/483601670685720591/1133893473469206578) `#help` to see if I can get a hand on this.

leepfrog commented:
Removed support for ember-3/ember-4 in the test-app and fixed a bug in the PageTitleService.  

All of the tests are now running locally for me, so..  maybe this time it'll work? :)

leepfrog commented:
There's a build error when using `CI=true` -- going to look into that and hopefully that'll be the last thing!

knownasilya commented:
Running test, lets see what happens

leepfrog commented:
There were enough small options / configuration changes that it made sense to just regenerate the test-app from the new blueprint and move over the tests.  So.. did that.

I also spun up into an ubuntu docker image and ran with the same `CI=true yarn test:ember` -- this works now!

leepfrog commented:
I didn't realize the CI was also running on my local fork 😅 -- but I was able to use it to get all of the steps green!  Should be good to re-run now.
